Add XModem feature

arl opened the pull request:
**UI additions**
- X/Y/ZModem buttons
- Disconnect button

**Xmodem transfer protocol implementation**
- with a progress dialog, cancellable

arl committed:
add buttons (disconnect, x/y/zmodem)

arl committed:
add and emit sessionStopped signal from SessionManager

arl committed:
enable/disable UI elements regarding session state

refactoring :

instead of having sessionStarted/Stopped signals handled :
    - in some cases, some slots connected to UI elements directly
    - in other cases, a specially made method because signal/slots
      parameter don't match

i refactored to handle each signal at one place only

also renamed transfer buttons

arl committed:
intermediary commit

arl committed:
handle x/y/z transfer buttons

arl committed:
rename FileTransferMode enum

arl committed:
add TransferThread class

arl committed:
add xmodem implementation

xmodem implementation (BSD licence)

code comes from https://github.com/kelvinlawson/xmodem-1k
and is a slightly modified version of Georges Menie's Xmodem
implementation
this code supports XModem-1K variant

arl committed:
add XModemTransfer class

XModemTransfer class:
 - inherits from the abstract TransferThread class
 - implements the startTransfer virtual method
 - encaspulates calls to the xmodem implementation code
 - always check port is open before open/write

sessionmanager :
 - fix crashes due to serialport not being in the same thread as
   worker
 - correctly delete thread
 - no more needed to keep a FileTransfer member instance
 - error signals emitted correctly from worker thread

transferthread :
 - renamed class into Filetransfer

arl committed:
rename xmodemtransfer to filetransfer

arl committed:
manage transfer stop (at protocol level)

xmodem cancel:

- added a _quit global variable to represent the fact that the transfer
should stop asap
- FileTransfer::Canceltransfer is now pure virtual, its implementation
will depend of the protocol
- remove timeout at Filetransfer level (unused)
- comments on the global variables

arl committed:
add transfer progress dialog

FileTransfer
- record total size and currently transferred byte amount, and emit a
signal every 1% percent increase

SessionManager
- add timer that updates progress dialog
- timer is handled in FileTransfer, because knowing the amount of
transferred byte is protocol dependent

arl committed:
code polishing, function renaming, comments, etc.

arl committed:
add CREDITS file

gvaldambrini commented on cutecom-ng.pro:
```diff
@@ -19,7 +19,11 @@ SOURCES += main.cpp\
-    searchhighlighter.cpp
+    searchhighlighter.cpp \
+    xmodem.cpp \
+    crc16.cpp \
```
potrebbe aver senso mettere le librerie di terze parti in un'altra directory, per esempio `libs`.

gvaldambrini commented on filetransfer.cpp:
```diff
@@ -0,0 +1,85 @@
+
+FileTransfer::~FileTransfer()
+{
+}
```
Quale è il senso di questo distruttore vuoto?
HINT: me l'ha generato Qt Creator non è una scusa valida.

gvaldambrini commented on filetransfer.h:
```diff
@@ -0,0 +1,134 @@
+
+/**
+ * \brief serial port file transfer
+ */
```
Poiché la classe FileTransfer contiene dei metodi virtuali puri, sarebbe il caso di documentare quali sono e che cosa dovrebbero fare.

gvaldambrini commented on filetransfer.cpp:
```diff
@@ -0,0 +1,85 @@
+            buffer = file.readAll();
+            total_size = file.size();
+        }
+    }
```
Uhm, a che serve questo blocco? E che succede se la open sul file fallisce?

gvaldambrini commented on filetransfer.cpp:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    // call child class file transfer protocol implementation
+    TransferError ret = this->performTransfer();
```
`this`? perché lo usi?

arl committed:
create libs folder and move xmodem files in it

arl committed:
remove not needed empty dtor

arl committed:
pure virtual methods documentation

arl committed:
treat FileTransfer input file error

arl committed:
remove unecessary this keyword

arl committed:
switch x/y/z buttons for combobox

arl committed:
rename FileTransferMode to Protocol for clarity

arl committed:
better 'Cancel Transfer' button handling

gvaldambrini commented on filetransfer.cpp:
```diff
@@ -0,0 +1,83 @@
+ * Copyright 2015 Develer S.r.l. (http://www.develer.com/)
+ * -->
+ *
+ * \brief TransferThread class implementation
```
`FileTransfer`, non `TransferThread` :D

gvaldambrini commented on filetransfer.cpp:
```diff
@@ -0,0 +1,83 @@
+ *
+ * \brief TransferThread class implementation
+ *
+ * \author Aurelien Rainone <aurelien@develer.org>
```
correggi la tua mail, è `@develer.com`, `@develer.org` per la email viene considerata spam dal nostro server di posta.

gvaldambrini commented on filetransfer.h:
```diff
@@ -0,0 +1,150 @@
+ */
+
+#ifndef TRANSFERTHREAD_H
+#define TRANSFERTHREAD_H
```
anche qua sono da aggiornare le guardie di compilazione.

arl committed:
replace all occurences of TransferThread

arl committed:
replace email in file headers

arl committed:
remove not needed dtor

arl committed:
rename initFileTransfer() -> transferFile()

arl committed:
remove unneeded this keyword

arl committed:
add forgotten emit keyword

arl committed:
remove unnecessary includes

arl committed:
FileTransfer handles its thread all by itself

goal : simpler structure and separation of concerns. SessionManager
doesn't have to be aware of thread management that is specific to
FileTransfer.

FileTransfer::startTransfer() is now non-blocking, it starts the thread and returns immediately; it is now the pure virtual method FileTransfer::startTransfer() that must emit

arl committed:
move progress timer management into FileTransfer

arl committed:
remove timer, move progress dialog to mainwindow

- forward FileTransfer signals from SessionManager to MainWindow
- removed majority of Qt::DirectConnection connect calls, but one :
at sessionmanager.cpp:217

    connect(this, &SessionManager::progressDialogCancelled,
            file_transfer, &FileTransfer::cancelTransfer,
            Qt::DirectConnection);
The fact that FileTransfer uses QThread is an implementation detail of
FileTransfer. The thread is actually started when and only when the file
transfer protocol begins. But at the time and space of the signal/slot
connection, we don't know if the transfer will really take place (chosen
file could be empty, or not readable for example).

arl committed:
correct bug in MainWindow::handleFileTransferProgressed

arl committed:
disconnect and reconnect serial port error handling while transferring

arl committed:
remove unneeded include

minor modif : comments, etc.

arl committed:
minor modifications : comments, whitespaces, etc.

aurelien-rainone commented:
updated PR

gvaldambrini commented on sessionmanager.h:
```diff
@@ -88,6 +98,48 @@ class SessionManager : public QObject
+     * \brief signal forwarded from outside to indicate that current
+     * file transfer should be cancelled
+     */
+    void transferRequestedByUser();
```
ma non dovrebbe chiamarsi transferCancelledByUser?

gvaldambrini commented on filetransfer.cpp:
```diff
@@ -0,0 +1,94 @@
+    // move serial instance back to main thread
+    serial->moveToThread(QApplication::instance()->thread());
+    thread->quit();
+    this->deleteLater();
```
Mi piace un po' poco questo _suicidio_ della classe `FileTransfer`. Tutta la gestione del thread è fatta al suo interno, quindi mi torna che debba terminare il thread qua dentro, tuttavia potevi connettere il segnale `handleTransferEnded` ad uno slot della classe `SessionManager` nel quale:
- emettere a sua volta il segnale `fileTransferEnded`
- chiamare la deleteLater sulla `FileTransfer`. 
  Ti dovrai memorizzare il puntatore alla `FileTransfer` o usare il `sender()` ma mi sembra una soluzione più pulita. C'è qualche ragione per cui non l'hai fatto così?

gvaldambrini commented on sessionmanager.cpp:
```diff
@@ -6,20 +6,23 @@
+#include <QProgressDialog>
+#include <QMessageBox>
 #include <QFile>
+#include <QTimer>
```
Vedo che hai eliminato la necessità di usare un timer, ma l'include l'hai lasciato nel caso cambiassi idea? :dancers:

arl committed:
method name typo : transferCancelledByUser

arl committed:
no more FileTransfer suicide, deletion done in SessionManager

arl committed:
remove unneeded #include QTimer

aurelien-rainone commented:
updated PR

arl committed:
transfer cancellation across threads with volatile

instead of using Qt signals/slots mechanism, that in this specific case
adds no benefits, and to remove the Qt::DirectConnection used
to receive a signal (accross threads) in a thread without event loop,
the solution of using a volatile flag was chosen and implemented

aurelien-rainone commented:
updated PR

arl committed:
remove duplicate connection

arl committed:
FileTransfer::quit_requested as public

arl committed:
use postEvent(QEvent::DeferredDelete) instead deleteLater

arl committed:
move file_transfer back to main thread
